core/libp2p/muxer/yamux/yamux_frame.cpp

```diff
@@ -0,0 +1,134 @@
+  }
+
+  kagome::common::Buffer pingOutMsg(uint32_t value) {
+    return YamuxFrame::frameBytes(YamuxFrame::kDefaultVersion,
+                                  YamuxFrame::FrameType::PING,
+                                  YamuxFrame::Flag::SYN, 0, value);
+  }
+
+  kagome::common::Buffer pingResponseMsg(uint32_t value) {
+    return YamuxFrame::frameBytes(YamuxFrame::kDefaultVersion,
+                                  YamuxFrame::FrameType::PING,
+                                  YamuxFrame::Flag::ACK, 0, value);
+  }
+
+  kagome::common::Buffer dataMsg(YamuxFrame::StreamId stream_id,
+                                 const kagome::common::Buffer &data) {
+    return YamuxFrame::frameBytes(YamuxFrame::kDefaultVersion,
+                                  YamuxFrame::FrameType::DATA,
+                                  YamuxFrame::Flag::SYN, stream_id,
+                                  static_cast<uint32_t>(data.size()), data);
+  }
+
+  kagome::common::Buffer goAwayMsg(YamuxFrame::GoAwayError error) {
+    return YamuxFrame::frameBytes(
+        YamuxFrame::kDefaultVersion, YamuxFrame::FrameType::GO_AWAY,
+        YamuxFrame::Flag::SYN, 0, static_cast<uint32_t>(error));
+  }
+
+  std::optional<YamuxFrame> parseFrame(gsl::span<const uint8_t> frame_bytes) {
+    if (frame_bytes.size() < YamuxFrame::kHeaderLength) {
+      return {};
+    }
+
+    YamuxFrame frame{};
+
+    frame.version_ = frame_bytes[0];
+
+    switch (frame_bytes[1]) {
+      case 0:
+        frame.type_ = YamuxFrame::FrameType::DATA;
+        break;
+      case 1:
+        frame.type_ = YamuxFrame::FrameType::WINDOW_UPDATE;
+        break;
+      case 2:
+        frame.type_ = YamuxFrame::FrameType::PING;
+        break;
+      case 3:
+        frame.type_ = YamuxFrame::FrameType::GO_AWAY;
+        break;
+      default:
+        return {};
+    }
+
+    switch ((static_cast<uint16_t>(frame_bytes[3]) << 8) | frame_bytes[2]) {
+      case 1:
+        frame.flag_ = YamuxFrame::Flag::SYN;
+        break;
+      case 2:
+        frame.flag_ = YamuxFrame::Flag::ACK;
+        break;
+      case 4:
+        frame.flag_ = YamuxFrame::Flag::FIN;
+        break;
+      case 8:
+        frame.flag_ = YamuxFrame::Flag::RST;
+        break;
+      default:
+        return {};
+    }
+
+    frame.stream_id_ = (static_cast<uint32_t>(frame_bytes[7]) << 24)
+        | (static_cast<uint16_t>(frame_bytes[6]) << 16)
+        | (static_cast<uint16_t>(frame_bytes[5]) << 8)
+        | (static_cast<uint16_t>(frame_bytes[4]));
+
+    frame.length_ = (static_cast<uint32_t>(frame_bytes[11]) << 24)
+        | (static_cast<uint16_t>(frame_bytes[10]) << 16)
+        | (static_cast<uint16_t>(frame_bytes[9]) << 8)
+        | (static_cast<uint16_t>(frame_bytes[8]));
+
+    const auto &data_begin = frame_bytes.begin() + YamuxFrame::kHeaderLength;
+    if (data_begin != frame_bytes.end()) {
+      frame.data_ = kagome::common::Buffer{
+          std::vector<uint8_t>(data_begin, frame_bytes.end())};
```

Comment:
also this function, stream would be SO MUCH better

Reply:
Also a ToDo